File: hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/hap/sign/SignElf.java
/*
 * Copyright (c) 2023-2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.hap.sign;

import com.ohos.hapsigntool.codesigning.exception.CodeSignException;
import com.ohos.hapsigntool.codesigning.exception.FsVerityDigestException;
import com.ohos.hapsigntool.codesigning.sign.CodeSigning;
import com.ohos.hapsigntool.hap.config.SignerConfig;
import com.ohos.hapsigntool.hap.entity.HwBlockHead;
import com.ohos.hapsigntool.hap.entity.HwSignHead;
import com.ohos.hapsigntool.hap.entity.SignBlockData;
import com.ohos.hapsigntool.hap.entity.SignatureBlockTags;
import com.ohos.hapsigntool.hap.entity.SignatureBlockTypes;
import com.ohos.hapsigntool.hap.exception.HapFormatException;
import com.ohos.hapsigntool.hap.exception.ProfileException;
import com.ohos.hapsigntool.utils.FileUtils;
import com.ohos.hapsigntool.utils.ParamConstants;
import com.ohos.hapsigntool.utils.ParamProcessUtil;
import com.ohos.hapsigntool.utils.StringUtils;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.DataOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * LiteOS bin file Signature signer.
 *
 * @since 2021/12/21
 */
public class SignElf {
    private static final Logger LOGGER = LogManager.getLogger(SignElf.class);

    private static final String CODESIGN_OFF = "0";

    private static final char CODESIGN_BLOCK_TYPE = 3;

    private static int blockNum = 0;

    /**
     * Constructor of Method
     */
    private SignElf() {
    }

    /**
     * Sign the bin file.
     *
     * @param signerConfig Config of the bin file to be signed.
     * @param signParams The input parameters of sign bin.
     * @return true if sign successfully; false otherwise.
     */
    public static boolean sign(SignerConfig signerConfig, Map<String, String> signParams) {
        boolean isSuccess = false;
        /* 1. Make block head, write to output file. */
        String inputFile = signParams.get(ParamConstants.PARAM_BASIC_INPUT_FILE);
        String outputFile = signParams.get(ParamConstants.PARAM_BASIC_OUTPUT_FILE);
        String profileSigned = signParams.get(ParamConstants.PARAM_BASIC_PROFILE_SIGNED);
        if (!writeBlockDataToFile(signerConfig, inputFile, outputFile, profileSigned, signParams)) {
            LOGGER.error("The block head data made failed.");
            ParamProcessUtil.delDir(new File(outputFile));
            return false;
        }
        LOGGER.info("The block head data made success.");

        /* 2. Make sign data, and write to output file */
        if (!writeSignHeadDataToOutputFile(inputFile, outputFile, blockNum)) {
            LOGGER.error("The sign head data made failed.");
            ParamProcessUtil.delDir(new File(outputFile));
        } else {
            isSuccess = true;
        }
        return isSuccess;
    }

    private static boolean writeBlockDataToFile(SignerConfig signerConfig,
        String inputFile, String outputFile, String profileSigned, Map<String, String> signParams) {
        try {
            String profileFile = signParams.get(ParamConstants.PARAM_BASIC_PROFILE);

            List<SignBlockData> signDataList = new ArrayList<>();

            long binFileLen = FileUtils.getFileLen(inputFile);
            if (binFileLen == -1) {
                LOGGER.error("file length is invalid, bin file len: " + binFileLen);
                throw new IOException();
            }
            // 1. generate sign data
            if (!StringUtils.isEmpty(signParams.get(ParamConstants.PARAM_BASIC_PROFILE))) {
                signDataList.add(generateProfileSignByte(profileFile, profileSigned));
            }
            blockNum = signDataList.size();
            SignBlockData codeSign = generateCodeSignByte(signerConfig, signParams, inputFile, blockNum, binFileLen);
            if (codeSign != null) {
                signDataList.add(0, codeSign);
            }
            blockNum = signDataList.size();
            // 2. use sign data generate offset and sign block head
            generateSignBlockHead(signDataList);

            return writeSignedElf(inputFile, signDataList, outputFile);
        } catch (IOException e) {
            LOGGER.error("writeBlockDataToFile failed.", e);
            return false;
        } catch (FsVerityDigestException | CodeSignException | HapFormatException | ProfileException e) {
            LOGGER.error("codesign failed.", e);
            return false;
        }
    }

    private static boolean writeSignedElf(String inputFile, List<SignBlockData> signBlockList, String outputFile) {
        try (FileOutputStream fileOutputStream = new FileOutputStream(outputFile);
             DataOutputStream dataOutputStream = new DataOutputStream(fileOutputStream)) {
            // 1. write the input file to the output file.
            if (!FileUtils.writeFileToDos(inputFile, dataOutputStream)) {
                LOGGER.error("Failed to write information of input file: " + inputFile
                        + " to outputFile: " + outputFile);
                throw new IOException();
            }

            // 2. write block head to the output file.
            for (SignBlockData signBlockData : signBlockList) {
                if (!FileUtils.writeByteToDos(signBlockData.getBlockHead(), dataOutputStream)) {
                    LOGGER.error("Failed to write Block Head to output file: " + outputFile);
                    throw new IOException();
                }
            }

            // 3. write block data to the output file.
            for (SignBlockData signBlockData : signBlockList) {
                boolean isSuccess;
                if (signBlockData.isByte()) {
                    isSuccess = FileUtils.writeByteToDos(signBlockData.getSignData(), dataOutputStream);
                } else {
                    isSuccess = FileUtils.writeFileToDos(signBlockData.getSignFile(), dataOutputStream);
                }

                if (!isSuccess) {
                    LOGGER.error("Failed to write Block Data to output file: " + outputFile);
                    throw new IOException();
                }
            }
        } catch (IOException e) {
            LOGGER.error("writeSignedBin failed.", e);
            return false;
        }
        return true;
    }

    private static void generateSignBlockHead(List<SignBlockData> signDataList)
            throws IOException {
        long offset = (long) HwBlockHead.getElfBlockLen() * signDataList.size();

        for (int i = 0; i < signDataList.size(); i++) {
            SignBlockData signBlockData = signDataList.get(i);

            signBlockData.setBlockHead(HwBlockHead.getBlockHeadLittleEndian(signBlockData.getType(),
                    SignatureBlockTags.DEFAULT, (int) signBlockData.getLen(), (int) offset));
            offset += signBlockData.getLen();
            if (isLongOverflowInteger(offset)) {
                LOGGER.error("The sign block " + i + "offset is overflow integer, offset: " + offset);
                throw new IOException();
            }
        }
    }

    private static SignBlockData generateProfileSignByte(String profileFile, String profileSigned) throws IOException {
        long profileDataLen = FileUtils.getFileLen(profileFile);

        if (profileDataLen == -1 || isLongOverflowShort(profileDataLen)) {
            LOGGER.error("file length is invalid, profileDataLen: " + profileDataLen);
            throw new IOException();
        }

        char isSigned = SignatureBlockTypes.getProfileBlockTypes(profileSigned);
        return new SignBlockData(profileFile, isSigned);
    }

    private static SignBlockData generateCodeSignByte(SignerConfig signerConfig, Map<String, String> signParams,
        String inputFile, int blockNum, long binFileLen) throws IOException,
            FsVerityDigestException, CodeSignException, HapFormatException, ProfileException {
        if (CODESIGN_OFF.equals(signParams.get(ParamConstants.PARAM_SIGN_CODE))) {
            return null;
        }
        CodeSigning codeSigning = new CodeSigning(signerConfig);
        long offset = binFileLen + (long) HwBlockHead.getElfBlockLen() * blockNum;
        String profileContent = signParams.get(ParamConstants.PARAM_PROFILE_JSON_CONTENT);
        byte[] codesignData = codeSigning.getCodeSignBlock(new File(inputFile), offset,
                signParams.get(ParamConstants.PARAM_IN_FORM), profileContent);
        return new SignBlockData(codesignData, CODESIGN_BLOCK_TYPE);
    }

    private static boolean writeSignHeadDataToOutputFile(String inputFile, String outputFile, int blockNum) {
        long size = FileUtils.getFileLen(outputFile) - FileUtils.getFileLen(inputFile);
        if (isLongOverflowInteger(size)) {
            LOGGER.error("File size is Overflow integer range.");
            return false;
        }
        HwSignHead signHeadData = new HwSignHead();
        byte[] signHeadByte = signHeadData.getSignHeadLittleEndian((int) size, blockNum);
        if (signHeadByte == null) {
            LOGGER.error("Failed to get sign head data.");
            return false;
        }
        return FileUtils.writeByteToOutFile(signHeadByte, outputFile);
    }

    private static boolean isLongOverflowInteger(long num) {
        return (num - (num & 0xffffffffL)) != 0;
    }

    private static boolean isLongOverflowShort(long num) {
        return (num - (num & 0xffffL)) != 0;
    }
}

File: pc/services/hap/sign/src/sign_elf.cpp
/*
 * Copyright (c) 2024-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "sign_elf.h"
#include "string_utils.h"
#include "code_signing.h"
#include "param_constants.h"
#include "hw_block_head.h"
#include "hw_sign_head.h"
#include "signature_block_types.h"
#include "signature_block_tags.h"

namespace OHOS {
namespace SignatureTools {

int SignElf::blockNum = 0;
const int SignElf::PAGE_SIZE = 4096;
const int SignElf::FILE_BUFFER_BLOCK = 16384;
const std::string SignElf::CODESIGN_OFF = "0";

bool SignElf::Sign(SignerConfig signerConfig, std::map<std::string, std::string> signParams)
{
    std::string inputFile = signParams.at(ParamConstants::PARAM_BASIC_INPUT_FILE);
    std::string tmpFile;
    bool checkAlignFileBy4kBytesFlag = AlignFileBy4kBytes(inputFile, tmpFile);
    if (!checkAlignFileBy4kBytesFlag) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] alignFileBy4kBytes error.");
        remove(tmpFile.c_str());
        return false;
    }
    std::string outputFile = signParams.at(ParamConstants::PARAM_BASIC_OUTPUT_FILE);
    std::string profileSigned = signParams.at(ParamConstants::PARAM_BASIC_PROFILE_SIGNED);
    bool checkWriteBlockDataToFileFlag = WriteBlockDataToFile(signerConfig, tmpFile,
                                                              outputFile, profileSigned, signParams);
    if (!checkWriteBlockDataToFileFlag) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] writeBlockDataToFile error.");
        remove(tmpFile.c_str());
        return false;
    }
    bool checkWriteSignHeadDataToOutputFileFlag = WriteSignHeadDataToOutputFile(tmpFile, outputFile, blockNum);
    if (!checkWriteSignHeadDataToOutputFileFlag) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] The sign head data made failed.");
        remove(tmpFile.c_str());
        return false;
    }
    return (remove(tmpFile.c_str()) == 0);
    ;
}

bool SignElf::AlignFileBy4kBytes(std::string& inputFile, std::string& ret)
{
    auto now = std::chrono::system_clock::now();
    auto timestamp = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
    std::string tmp = "tmpFile" + std::to_string(timestamp);
    std::ofstream output(tmp);
    bool checkoutputFlag = output.is_open();
    if (!checkoutputFlag) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] unable to open " + tmp + "s.");
        return false;
    }
    std::ifstream input(inputFile);
    bool checkinputFlag = input.is_open();
    if (!checkinputFlag) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] unable to open " + inputFile + "s.");
        return false;
    }
    char buffer[FILE_BUFFER_BLOCK];
    std::streamsize bytesRead;
    int64_t output_length = 0;
    while ((bytesRead = input.read(buffer, sizeof(buffer)).gcount()) > 0) {
        output.write(buffer, bytesRead);
        output_length += bytesRead;
    }
    int64_t addLength = PAGE_SIZE - (output_length % PAGE_SIZE);
    std::vector<char> bytes(addLength, 0);
    output.write(bytes.data(), addLength);
    ret = tmp;
    return true;
}

bool SignElf::WriteBlockDataToFile(SignerConfig signerConfig,
                                   std::string inputFile, std::string outputFile, std::string profileSigned,
                                   std::map<std::string, std::string> signParams)
{
    std::string profiliFile = signParams.at(ParamConstants::PARAM_BASIC_PROFILE);
    std::list<SignBlockData> signDataList;
    int64_t binFileLen = FileUtils::GetFileLen(inputFile);
    bool checkFlag = binFileLen < 0 || IsLongOverflowInteger(binFileLen);
    if (checkFlag) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] getFileLen or isLongOverflowInteger error.");
        return false;
    }
    bool checkIsEmptyFlag = StringUtils::IsEmpty(signParams.at(ParamConstants::PARAM_BASIC_PROFILE));
    if (!checkIsEmptyFlag) {
        signDataList.push_front(GenerateProfileSignByte(profiliFile, profileSigned));
    }
    blockNum = signDataList.size() + 1;
    SignBlockData* codeSign = nullptr;
    bool checkGenerateCodeSignByteFlag = !GenerateCodeSignByte(signerConfig, signParams, inputFile, blockNum,
                                                               binFileLen, &codeSign) || !codeSign;
    if (checkGenerateCodeSignByteFlag) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] generateCodeSignByte error.");
        if (codeSign) {
            delete codeSign;
        }
        return false;
    }
    signDataList.push_front(*codeSign);
    blockNum = signDataList.size();
    bool checkGenerateSignBlockHeadFlag = GenerateSignBlockHead(signDataList);
    if (!checkGenerateSignBlockHeadFlag) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] generateSignBlockHead error.");
        delete codeSign;
        return false;
    }
    delete codeSign;
    return WriteSignedElf(inputFile, signDataList, outputFile);
}

bool SignElf::WriteSignedElf(std::string inputFile, std::list<SignBlockData>& signBlockList, std::string outputFile)
{
    std::ifstream fileInputStream(inputFile, std::ios::binary);
    std::ofstream fileOutputStream(outputFile, std::ios::binary);
    bool checkFlag = !fileInputStream.is_open() || !fileOutputStream.is_open();
    if (checkFlag) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] inputFile or outputFile open error.");
        return false;
    }
    char buffer[FILE_BUFFER_BLOCK];
    while (!fileInputStream.eof()) {
        fileInputStream.read(buffer, sizeof(buffer));
        fileOutputStream.write(buffer, fileInputStream.gcount());
    }
    for (auto signBlockData : signBlockList) {
        bool checkWriteByteToOutFileFlag = FileUtils::WriteByteToOutFile(signBlockData.GetBlockHead(),
                                                                         fileOutputStream);
        if (!checkWriteByteToOutFileFlag) {
            PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "Failed to write Block Head to output file.");
            fileInputStream.close();
            fileOutputStream.close();
            return false;
        }
    }
    for (auto signBlockData : signBlockList) {
        bool isSuccess;
        checkFlag = signBlockData.GetByte();
        if (checkFlag) {
            isSuccess = FileUtils::WriteByteToOutFile(signBlockData.GetSignData(), fileOutputStream);
        } else {
            std::ifstream InputSignFileStream(signBlockData.GetSignFile(), std::ios::binary);
            bool checkFileFlag = !InputSignFileStream.is_open();
            if (checkFileFlag) {
                PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] InputSignFileStream open error.");
                return false;
            }
            int result = FileUtils::WriteInputToOutPut(InputSignFileStream, fileOutputStream,
                                                       (long)signBlockData.GetLen());
            isSuccess = (result == 0 ? true : false);
        }
        if (!isSuccess) {
            PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "Failed to WriteByteToOutFile.");
            fileInputStream.close();
            fileOutputStream.close();
            return false;
        }
    }
    fileInputStream.close();
    fileOutputStream.close();
    return true;
}

bool SignElf::GenerateSignBlockHead(std::list<SignBlockData>& signDataList)
{
    int64_t offset = HwBlockHead::GetElfBlockLen() * signDataList.size();
    for (std::list<SignBlockData>::iterator it = signDataList.begin(); it != signDataList.end(); ++it) {
        std::vector<int8_t> tmp = HwBlockHead::GetBlockHeadLittleEndian(it->GetType(),
                                                                        SignatureBlockTags::DEFAULT,
                                                                        it->GetLen(), offset);
        it->SetBlockHead(tmp);
        offset += it->GetLen();
        bool checkIsLongOverflowIntegerFlag = IsLongOverflowInteger(offset);
        if (checkIsLongOverflowIntegerFlag) {
            PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] The sign block offset is overflow integer.");
            return false;
        }
    }
    return true;
}

SignBlockData SignElf::GenerateProfileSignByte(std::string profileFile, std::string profileSigned)
{
    int64_t profileDataLen = FileUtils::GetFileLen(profileFile);
    if (profileDataLen < 0 || IsLongOverflowShort(profileDataLen)) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] generateProfileSignByte profileDataLen error.");
    }
    char isSigned = SignatureBlockTypes::GetProfileBlockTypes(profileSigned);
    return SignBlockData(profileFile, isSigned);
}

bool SignElf::GenerateCodeSignByte(SignerConfig signerConfig, std::map<std::string, std::string> signParams,
                                   std::string inputFile, int blockNum, long binFileLen, SignBlockData** codeSign)
{
    if (signParams.at(ParamConstants::PARAM_SIGN_CODE) == CODESIGN_OFF) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] generateCodeSignByte PARAM_SIGN_CODE error.");
        return false;
    }
    CodeSigning codeSigning(&signerConfig);
    long offset = binFileLen + (long)HwBlockHead::GetElfBlockLen() * blockNum;
    std::string profileContent = signParams.at(ParamConstants::PARAM_PROFILE_JSON_CONTENT);
    std::vector<int8_t> codesignData;
    bool checkGetElfCodeSignBlockFlag = codeSigning.GetElfCodeSignBlock(inputFile, offset,
                                                                        signParams.at(ParamConstants::PARAM_IN_FORM),
                                                                        profileContent, codesignData);
    if (!checkGetElfCodeSignBlockFlag) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR, "[SignElf] GetElfCodeSignBlock error.");
        return false;
    }
    *codeSign = new SignBlockData(codesignData, CODESIGN_BLOCK_TYPE);
    return true;
}

bool SignElf::WriteSignHeadDataToOutputFile(std::string inputFile, std::string outputFile, int blockNum)
{
    int64_t size = FileUtils::GetFileLen(outputFile) - FileUtils::GetFileLen(inputFile);
    if (IsLongOverflowInteger(size)) {
        PrintErrorNumberMsg("SIGN_ERROR", SIGN_ERROR,
                            "[SignElf] writeSignHeadDataToOutputFile isLongOverflowInteger error.");
        return false;
    }
    HwSignHead signHeadData;
    std::vector<int8_t> signHeadByte = signHeadData.GetSignHeadLittleEndian((int)size, blockNum);
    std::ofstream fileOutputStream(outputFile, std::ios::app | std::ios::binary);
    return FileUtils::WriteByteToOutFile(signHeadByte, fileOutputStream);
}

bool SignElf::IsLongOverflowInteger(int64_t num)
{
    return (num - (num & 0xffffffffL)) != 0;
}

bool SignElf::IsLongOverflowShort(int64_t num)
{
    return (num - (num & 0xffffL)) != 0;
}

} // namespace SignatureTools
} // namespace OHOS